server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -941,35 +916,33 @@ private static Metadata metadataForSnapshot(SnapshotsInProgress.Entry snapshot,
         String repository,
         List<String> snapshots
     ) {
-        if (snapshotsInProgress == null || snapshotsInProgress.entries().isEmpty()) {
+        if (snapshotsInProgress == null || snapshotsInProgress.isEmpty()) {
             return Collections.emptyList();
         }
         if ("_all".equals(repository)) {
-            return snapshotsInProgress.entries();
+            return snapshotsInProgress.asStream().collect(Collectors.toUnmodifiableList());
         }
-        if (snapshotsInProgress.entries().size() == 1) {
+        if (snapshotsInProgress.count() == 1) {
             // Most likely scenario - one snapshot is currently running
             // Check this snapshot against the query
-            SnapshotsInProgress.Entry entry = snapshotsInProgress.entries().get(0);
-            if (entry.snapshot().getRepository().equals(repository) == false) {
+            List<SnapshotsInProgress.Entry> inRepo = snapshotsInProgress.forRepo(repository);
+            if (inRepo.isEmpty()) {
                 return Collections.emptyList();
             }
-            if (snapshots.isEmpty() == false) {
+            if (snapshots.isEmpty()) {
+                return inRepo;
+            } else {
+                final SnapshotsInProgress.Entry entry = inRepo.get(0);
                 for (String snapshot : snapshots) {
                     if (entry.snapshot().getSnapshotId().getName().equals(snapshot)) {
-                        return snapshotsInProgress.entries();
+                        return inRepo;
                     }
                 }
                 return Collections.emptyList();
-            } else {
-                return snapshotsInProgress.entries();
             }
         }
         List<SnapshotsInProgress.Entry> builder = new ArrayList<>();
-        for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
-            if (entry.snapshot().getRepository().equals(repository) == false) {
-                continue;
-            }
+        for (SnapshotsInProgress.Entry entry : snapshotsInProgress.forRepo(repository)) {
             if (snapshots.isEmpty() == false) {
```

Comment:
If `snapshots.isEmpty()` can we always just return `snapshotsInProgress.forRepo(repository)`?

Reply:
Yea did this and dropped the count == 1 optimization as well.